%\VignetteIndexEntry{A short SINCE}
%\VignetteKeywords{Single Cell} 
%\VignettePackage{SINCE}

%
% NOTE -- ONLY EDIT THE .Rnw FILE!!!  The .tex file is
% likely to be overwritten.
%
\documentclass[11pt]{article}



\textwidth=6.2in
\textheight=8.5in
%\parskip=.3cm
\oddsidemargin=.1in
\evensidemargin=.1in
\headheight=-.3in

\newcommand{\scscst}{\scriptscriptstyle}
\newcommand{\scst}{\scriptstyle}


\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textit{#1}}}
\newcommand{\Rmethod}[1]{{\texttt{#1}}}
\newcommand{\Rfunarg}[1]{{\texttt{#1}}}
\newcommand{\Rclass}[1]{{\textit{#1}}}

\textwidth=6.2in

\bibliographystyle{plainnat} 
 
\begin{document}
%\setkeys{Gin}{width=0.55\textwidth}

\title{The \Rpackage{SINCE} package}
\author{Jie Hao, Xin Zou, Ze Guang Han}
\maketitle

% ---------------------------------------------------------------------------

\section{Introduction}

The main function performs SINCE to evaluate clustering results with different parameters.
\Rpackage{SINCE} package.


% ---------------------------------------------------------------------------

\section{Rpackage{SINCE} Analysis Example Workflow}

Required packages: knitr, Matrix, scater, SC3, SingleCellExperiment, OGFSC and SINCE.
install knitr, Matrix from cran.
install scater,SC3,SingleCellExperiment from Bioconductor.
install OGFSC: The OGFSC package is freely downloadable from github: https://github.com/XZouProjects/OGFSC-R

A demo script and the associated files can be found at: https://github.com/XZouProjects/SINCE

Brief introduction of SINCE:

The SINCE algorithm is designed to estimate the number of cell clusters by incorporating with existing clustering methods. In the demo script, the SC3 method is used to achieve cell assignment results with different numbers of clusters. Then SINCE is applied to evaluate each SC3 clustering result. Based on the SINCE output, i.e., CERS values, one can determine the optimal number of clusters.

Step by step tutorial:

1. Gene filtering by OGFSC.
2. SC3 clustering with various numbers of clusters. 
3. Save each clustering result into a separate file using the write.table() function.
4. Manually make up a clustering-result-directory list, and save the list into an independent file. For example, the file name is ``cluster\_res\_list.txt`` in the demo script.
5. Run SINCE using the OGFSC filtered data and the directory list file as compulsory inputs. 

SINCE parameters:

Compulsory input:
data - a log2 transformed scRNA-seq data matrix with rows being genes and columns being cells.
cluster\_res\_list - a .txt file containing the list of clustering results.

Sdjustable parameters:
CERS\_plot\_option - the option to show CERS plot. By default 1.
r2Cutoff - the OPLSDA r2 cutoff values in variable gene selection. By default[0.5, 0.7, 0.8].
paralSize - number of repeats when calculating CERS. By default 6.
miniClusterSize - the minimum size of each cell cluster. By default 5.

Example:

load demo data 
<<a1>>=
library(SINCE)
url = "https://raw.githubusercontent.com/XZouProjects/SINCE_test/master/SC3_Pollen_2.xls"
download.file(url, "SC3_Pollen_2.xls")
@

The clustering result directories are listed in the file named ``cluster\_res\_list.txt``
<<a3>>=
url = "https://raw.githubusercontent.com/XZouProjects/SINCE_test/master/cluster_res_list.txt"
download.file(url,"cluster_res_list.txt")
@

<<a4>>=
url = "https://raw.githubusercontent.com/XZouProjects/SINCE_test/master/data.rda"
download.file(url, "data.rda")
load("data.rda")
@

SINCE analysis
<<a2>>=
CERS <- SINCE(data = data, cluster_res_list = "cluster_res_list.txt", paralSize = 4)
@
% ---------------------------------------------------------------------------


\end{document}
